### PR TITLE
PA: 1155, PA: 212 - Get Data service PVC to full condition and resize 

### PR DESCRIPTION
**Pull request by jkagliwal-px**
This PR is with respect to two JIRA tickets -
1. https://portworx.atlassian.net/browse/PA-1155 (Create a function to poll for PVC usage and resize after full usage)
2. https://portworx.atlassian.net/browse/PA-212 (testcase - Get Data service PVC to full condition and resize the pvc)

Enhancements made -> 
1.  Used portworx libraries to get the volumes and inspect the usage of the pvc.
2. After reaching a threshold, resizing the pvc by 1 gb.
3. To fill the pvc, pds workload scripts are executed.
4. Introduced a new function to pass custom resource templates from the testcase.

PS: 1.This testcase currently runs only for Postgres as I need to pass a custom resource template with a small size storage for faster test execution.

Jenkins Run - 
https://jenkins.pwx.dev.purestorage.com/job/PDS/job/pds-branch-build-test/358/console

### CODE REVIEW

File: tests/pds/common.go [hunk 1]
```diff
@@ -162,3 +167,84 @@ func RunWorkloads(params pdslib.WorkloadGenerationParams, ds PDSDataService, dep
 	return pod, dep, err
 
 }
+
+// Check the DS related PV usage and resize in case of 90% full
+func CheckPVCtoFullCondition(context []*scheduler.Context) error {
+	log.Infof("Start polling for the pvc consumption by the DS")
```
Comment (Madanagopal19) [inline tests/pds/common.go:173]: nit: update the logger like "check pvc usage"

File: tests/pds/common.go [hunk 2]
```diff
@@ -162,3 +167,84 @@ func RunWorkloads(params pdslib.WorkloadGenerationParams, ds PDSDataService, dep
...
+			for _, vol := range vols {
+				appVol, err := tests.Inst().V.InspectVolume(vol.ID)
+				if err != nil {
+					return nil, true, err
+				}
+				pvcCapacity := appVol.Spec.Size / units.GiB
+				usedGiB := appVol.GetUsage() / units.GiB
+				threshold := pvcCapacity - 1
+				if usedGiB >= threshold {
+					log.Infof("Capacity in GB is %v", pvcCapacity)
```
Comment (Madanagopal19) [inline tests/pds/common.go:189]: nit: combine the possible info logs

File: tests/pds/common.go [hunk 3]
```diff
@@ -162,3 +167,84 @@ func RunWorkloads(params pdslib.WorkloadGenerationParams, ds PDSDataService, dep
...
+		return nil, true, fmt.Errorf("threshold not achieved for the PVC, ")
+	}
+	_, err := task.DoRetryWithTimeout(f, 30*time.Minute, 15*time.Second)
+
+	return err
+}
+
+// Increase PVC by 1 gb
+func IncreasePVCby1Gig(context []*scheduler.Context) error {
+	log.Info("Resizing of the PVC begins ...")
```
Comment (Madanagopal19) [inline tests/pds/common.go:206]: nit: remove the dots at the end of the log message

File: tests/pds/dataservice_test.go [hunk 4]
```diff
@@ -1666,3 +1666,67 @@ var _ = Describe("{AddAndValidateUserRoles}", func() {
...
+					}
+				})
+
+				Step("Checking the PVC usage", func() {
+					ctx := dsTest.CreateSchedulerContextForPDSApps(depList)
+					err = CheckPVCtoFullCondition(ctx)
+					log.FailOnError(err, "Failing while filling the PVC to 90 percentage of its capacity due to ...")
+					err = IncreasePVCby1Gig(ctx)
+					log.FailOnError(err, "Failing while Increasing the PVC name...")
+				})
```
Comment (Madanagopal19) [inline tests/pds/dataservice_test.go:1724]: check the deployment health status after resizing the pvc

File: tests/pds/dataservice_test.go [hunk 5]
```diff
@@ -1666,3 +1666,67 @@ var _ = Describe("{AddAndValidateUserRoles}", func() {
...
+					controlPlane.UpdateResourceTemplateName("pds-auto-pvcFullCondition")
+					deployment, _, dataServiceVersionBuildMap, err = DeployandValidateDataServices(ds, params.InfraToTest.Namespace, tenantID, projectID)
+					log.FailOnError(err, "Error while deploying data services")
+					deployments[ds] = deployment
+					dsVersions[ds.Name] = dataServiceVersionBuildMap
+					depList = append(depList, deployment)
+					dsName = ds.Name
+
+				})
+			}
```
Comment (Madanagopal19) [inline tests/pds/dataservice_test.go:1696]: add defer function to clean up the created deployment

File: tests/pds/dataservice_test.go [hunk 6]
```diff
@@ -1666,3 +1666,67 @@ var _ = Describe("{AddAndValidateUserRoles}", func() {
...
+							} else {
+								generateWorkloads[ds.Name] = dep.Name
+							}
+							for dsName, workloadContainer := range generateWorkloads {
+								log.Debugf("dsName %s, workloadContainer %s", dsName, workloadContainer)
+							}
+						}
+					}
+				})
+
```
Comment (Madanagopal19) [inline tests/pds/dataservice_test.go:1717]: Add defer function to clean up the created workload pods

File: tests/pds/dataservice_test.go [hunk 7]
```diff
@@ -1666,3 +1666,67 @@ var _ = Describe("{AddAndValidateUserRoles}", func() {
...
+			// This testcase is currently applicable only for postgresql deployments
+			if dsName == postgresql {
+				Step("Running Workloads before scaling up PVC ", func() {
+					for ds, deployment := range deployments {
+						if Contains(dataServicePodWorkloads, ds.Name) || Contains(dataServiceDeploymentWorkloads, ds.Name) {
+							log.InfoD("Running Workloads on DataService %v ", ds.Name)
+							var params pdslib.WorkloadGenerationParams
+							pod, dep, err = RunWorkloads(params, ds, deployment, namespace)
+							log.FailOnError(err, fmt.Sprintf("Error while genearating workloads for dataservice [%s]", ds.Name))
+							if dep == nil {
```
Comment (Madanagopal19) [inline tests/pds/dataservice_test.go:1706]: since we are running the test only for postgresql we can have only 
`generateWorkloads[ds.Name] = dep.Name`

**Review by Madanagopal19 — CHANGES_REQUESTED** [review]

File: tests/pds/common.go (hunk 1 above)
Comment (jkagliwal-px) [reply tests/pds/common.go:173]: Done !

File: tests/pds/common.go (hunk 2 above)
Comment (jkagliwal-px) [reply tests/pds/common.go:189]: Done !

File: tests/pds/common.go (hunk 3 above)
Comment (jkagliwal-px) [reply tests/pds/common.go:206]: Done!

File: tests/pds/dataservice_test.go (hunk 4 above)
Comment (jkagliwal-px) [reply tests/pds/dataservice_test.go:1724]: Done !

File: tests/pds/dataservice_test.go (hunk 5 above)
Comment (jkagliwal-px) [reply tests/pds/dataservice_test.go:1696]: Done !

File: tests/pds/dataservice_test.go (hunk 6 above)
Comment (jkagliwal-px) [reply tests/pds/dataservice_test.go:1717]: Done !

File: tests/pds/dataservice_test.go (hunk 7 above)
Comment (jkagliwal-px) [reply tests/pds/dataservice_test.go:1706]: Done!

File: drivers/pds/dataservice/dataservices.go [hunk 8]
```diff
@@ -456,6 +456,11 @@ func (d *DataserviceType) CreateSchedulerContextForPDSApps(pdsApps []*pds.Models
 	var ctx *scheduler.Context
 
 	for _, dep := range pdsApps {
+		dep.Namespace, err = components.Namespace.GetNamespace(*dep.NamespaceId)
+		if err != nil {
```
Comment (Madanagopal19) [inline drivers/pds/dataservice/dataservices.go:460]: return the err here

**Review by Madanagopal19 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**